Propagate location type changes automatically

esoergel opened the pull request:
https://github.com/dimagi/commcare-hq/pull/9946 is based off of https://github.com/dimagi/commcare-hq/pull/9860, so this is a PR of just the differences.  Be sure to check out comments on the original PR too.

http://manage.dimagi.com/default.asp?188796
Location types have an administrative property which determines whether or not to create a supply point case for that location. In the UI this is called "Tracks Stock", where tracks_stock == not administrative.
Currently, when you change a location from administrative to not administrative, nothing much happens - it only affects future changes. This PR makes that change propagate to all locations of that type. This happens asynchronously.
Also, we currently only create or update supply point cases when a location is changed via the corresponding django form. This resulted in a lot of confusing test code that had to manually create supply points, otherwise certain things wouldn't work. I moved that function into SQLLocation.save(), so whenever you update a location, it'll automatically update the supply point. This is a noop if it won't actually result in any changes to the supply point case.
Another issue I ran up against was that we have been defaulting all location types on non-commtrack domains to administrative=False, that is - creating supply point cases for them. I spoke with @sheelio and decided to stop this, although I imagine at some point we might want to migrate existing affected locations.
This is on staging right now and I'd like to do more testing before merge, but figured I'd open it up for review.
@czue

esoergel committed:
failing test

esoergel committed:
Make supply-points sync from SQLLocation.save()

esoergel committed:
Stop manually making supply points in tests

And also stop creating locations without domains

esoergel committed:
Fix tests

esoergel committed:
Add stock_tracking_types to setup_locs_and_types

esoergel committed:
Test the gamut for stock-tracking changes

esoergel committed:
Stop manually making supply point cases

(it happens automatically now)

esoergel committed:
Remove unused dbaccessor

esoergel committed:
Use location.linked_supply_point

It uses the link on the sqllocation - the dbaccessor will return closed
cases.

esoergel committed:
Don't make non-commtrack, non-admin locations

Commit:
Fixed ewsghana tests

Commit:
Fixed typo

esoergel committed:
Auto open and close supply-points on change

esoergel committed:
Don't make stock data on non-commtrack domain

esoergel committed:
Failing test for StockState changes

esoergel committed:
Synchronously update StockStates as appropriate

esoergel committed:
Make administrative updates async

esoergel committed:
Make generic `locking_task`

esoergel committed:
Add help text to checkbox.

esoergel committed:
Cleanup docstring [ci skip]

dimagimon commented on corehq/util/decorators.py:
```diff
@@ -62,3 +64,65 @@ def __enter__(self):
+    return ("{}-" + format_str).format(fn.__name__, **kwargs)
+
+
+def locking_task(unique_key, default_retry_delay=30, timeout=5*60, max_retries=3):
```
E226 missing whitespace around arithmetic operator

czue commented on corehq/apps/locations/models.py:
```diff
@@ -107,13 +111,20 @@ def _populate_stock_levels(self):
-        return super(LocationType, self).save(*args, **kwargs)
+        saved = super(LocationType, self).save(*args, **kwargs)
+
+        if self._administrative_old != self.administrative:
```
should we add `and self._administrative_old is not None` here to avoid spinning up celery tasks when locations are first created?

esoergel replied:
`self._administrative_old` isn't set until after the `super` call to `__init__`, so it will already be set to either True or False:
```python
In [1]: lt = LocationType(domain='ethan-commtrack', name='foo')  

In [2]: lt.administrative  
Out[2]: False
```
That means it won't register a difference unless `administrative` is changed outside of the constructor (which could happen).  Actually, what we could do is add `is_first_save = self.pk is None` right before the actual save call, then skip the task if that's `True`.

czue commented on corehq/apps/locations/tests/util.py:
```diff
@@ -62,27 +62,31 @@ def create_locations(locations, types, parent):
 
 
-def setup_locations_and_types(domain, location_types, locations):
+def setup_locations_and_types(domain, location_types, stock_tracking_types, locations):
```
nit: i think it'd be easier to reason with this function if location_types were specs that also included whether or not they were stock tracking. seems a little odd to have the separate params. also I think this will crash if you pass something into `stock_tracking_types` that isn't in `location_types`. 

it's just test code so seems fine to leave unless you feel like changing

esoergel replied:
I considered that, but that makes it more difficult to use in tests, and that's really where I want the intent to be obvious.

czue commented on corehq/apps/locations/tests/test_change_administrative_status.py:
```diff
@@ -0,0 +1,133 @@
+        self.assertIsNone(loc.supply_point_id, msg)
+
+    def test_change_to_track_stock(self):
+        self.assertHasSupplyPoint(self.boston)
```
nit: might be cleaner to just put this check into `setUp` rather than calling it at the beginning of several tests

czue replied:
(or a separate test - or just remove this line from this test)

esoergel replied:
I was thinking of this as trying to illustrate what's happening - "first it has a supply point, then you set it to administrative, then it doesn't have a supply point, then you set it back..."

It's definitely not necessary to have in each one, but I think it makes the intent more obvious.

esoergel replied:
actually, it's irrelevant for this particular test case, I'll remove it, lol.

czue commented on corehq/apps/locations/tests/test_change_administrative_status.py:
```diff
@@ -0,0 +1,133 @@
+        # tracking stock again should restore the stock states
+        self.city_type.administrative = False
+        self.city_type.save()
+        self.assertTrue(has_stock_states(self.boston))
```
:heart_decoration: these tests

esoergel replied:
:bow:

czue commented on corehq/apps/locations/tasks.py:
```diff
@@ -0,0 +1,35 @@
+
+
+class CouldNotAqcuireLock(Exception):
+    pass
```
I think this can be deleted

esoergel replied:
:+1:

czue commented on corehq/apps/locations/templates/locations/location_types.html:
```diff
@@ -108,7 +108,17 @@
+                                      data-content="{% blocktrans %}
+                                          Does this location keep inventory and track stock?
+                                          Note that changing this property will update all
+                                          locations of this level accordingly.
```
might want to get @sheelio to review this. the second sentence seems kind of vague/confusing to me

sheelio replied:
@esoergel 

Updated text:
"Do locations at this organization level keep inventory and track stock?"

esoergel replied:
Sounds good.  @sheelio, I'm not sure how specific I should be here, but basically what I want to convey is that changing this property will affect all locations of that type (ie use with caution).

esoergel replied:
Just saw that you had already replied - must've needed to refresh.  @sheelio What about the second part?

sheelio replied:
Didn't think we needed to clarify the second part.  If we do then would use this:
Note: Changes to this settings may stop existing locations at this level from collecting stock information.

esoergel replied:
That sounds kinda cryptic and scary - how about "Turning this off will stop existing locations at this level from collecting stock information"

czue commented on corehq/util/decorators.py:
```diff
@@ -62,3 +64,65 @@ def __enter__(self):
+
+
+def _get_unique_key(format_str, fn, *args, **kwargs):
+    # Sorry this is so magic
```
haha, you could at least try to explain it

esoergel replied:
*sigh*... alright

czue commented on corehq/util/decorators.py:
```diff
@@ -62,3 +64,65 @@ def __enter__(self):
+
+        greet.delay(joeshmoe)
+        # Locking key used would be "greet-joeshmoe@test.commcarehq.org-Dimagi"
+    """
```
neat! agree that @gcapalbo should review this

esoergel replied:
Thanks, I'm hoping this will be generally useful to have around, I'll send something out to team-commcare-hq once this PR is merged.

gcapalbo replied:
(left comments in https://github.com/dimagi/commcare-hq/pull/9946)

czue commented:
looks good! just had a few minor comments. happy to merge this if that's easier, or we can close this and  https://github.com/dimagi/commcare-hq/pull/9860... not sure how you want to manage getting this on the path to merged to master

esoergel committed:
Don't start task on first save.

esoergel committed:
PR feedback

esoergel commented:
@czue I'd like to merge both together, after I hear back from QA.  The most straightforward way to do that is probably https://github.com/dimagi/commcare-hq/pull/9946, since that's the branch with all the changes on it.  So in that case, we can close this and https://github.com/dimagi/commcare-hq/pull/9860 whenever you're happy with them.

esoergel committed:
Clarify serial_task

dimagimon commented on corehq/util/decorators.py:
```diff
@@ -62,3 +64,71 @@ def __enter__(self):
+    return ("{}-" + format_str).format(fn.__name__, **kwargs)
+
+
+def serial_task(unique_key, default_retry_delay=30, timeout=5*60, max_retries=3,
```
E226 missing whitespace around arithmetic operator

Commit:
Use FormExportDataSchema

esoergel committed:
Fix and clarify is_new_user

In the POST method, it was never actually set.

esoergel committed:
Remove get_domain_type and cleanup

esoergel committed:
Stop linking to commtrack domain registration

esoergel committed:
Remove make new commtrack domain feature

Commit:
CaseExportSchema

Commit:
allow for arbitrary number of schemas

Commit:
_merged_schema -> _merged_schemas

Commit:
test for merging case data schemas

Commit:
Use constant for group name

Commit:
define generate_schema_from_builds

Commit:
[ci skip] dimagimon

Commit:
Update domain gateway list to use postgres models

Commit:
Update add domain backend view to use postgres models

Commit:
Update edit domain backend view to use postgres models

TylerSheffels committed:
Fix spacing

czue committed:
Merge pull request #10045 from dimagi/filter-export-by-domain

Filter export by domain

czue committed:
Merge pull request #10020 from dimagi/abt-reports

Update Abt SMS report data source

czue committed:
Merge pull request #10024 from dimagi/rm-mgmt-cmds

Remove a bunch of old management commands

czue committed:
move kafka tests to the beginning of group

czue committed:
[ci skip] clean up comment/doc string

czue committed:
add last_modified to data sources

snopoke committed:
add docker bootstraping

czue committed:
[ci skip] imports

czue committed:
put local apps first

czue committed:
kill OldRoles, get on path to killing OldPermissions

snopoke committed:
add dockerhq script, update README

snopoke committed:
allow running arbitrary commands

snopoke committed:
add grunt-cli and uglify-js

snopoke committed:
add project name and ps

snopoke committed:
Merge pull request #10034 from dimagi/send-to-kafka

Send SQL form updates to ElasticSearch

czue committed:
unused import

czue committed:
only show dashboard tab to mobile workers if they have a custom role

czue committed:
coerce visibility checks to booleans

czue committed:
extract out shared logic to a function

czue committed:
Merge pull request #10054 from dimagi/increase-periodic-timing-assert

Increase the time before triggering a soft assert on periodic timing

Commit:
Move global gateway list to class-based view

Commit:
Reuse domain level create/edit gateway views to make class-based global backend views

snopoke committed:
Merge pull request #10049 from dimagi/comment

[ci skip] clean up comment/doc string

NoahCarnahan committed:
Update `SplitExportColumn.get_value()` doc string

NoahCarnahan committed:
Merge pull request #10050 from dimagi/last-modified-data-source

add last_modified to data sources

Commit:
Remove old unused view code

orangejenny committed:
Add .hqm-existing-controls

millerdev committed:
Merge pull request #10018 from dimagi/docker

docker for realz

Commit:
Split get_backend_classes util

czue committed:
add basic test for elastic CasePillow

czue committed:
extract function

czue committed:
missed a dependency

czue committed:
extract out the function that makes cases

czue committed:
add failing test outline for sql cases to elastic

czue committed:
add case sql topic back

dannyroberts committed:
Merge pull request #10051 from dimagi/kill-old-permissins

kill OldRoles, get on path to killing OldPermissions

czue committed:
working end to end test getting cases into kafka

Commit:
Fixed Care Pathways custom reports

czue committed:
[ci skip] imports

czue committed:
use ensure_index_deleted

TylerSheffels committed:
Merge pull request #10056 from dimagi/mm-upload-previews

Add .hqm-existing-controls

czue committed:
don't fail if no cases

benrudolph committed:
Merge pull request #10028 from dimagi/server-side-xpath-validation

Server side xpath validation

esoergel committed:
Merge pull request #10041 from lwyszomi/opm-changes

OPM - changes in report

benrudolph committed:
Merge pull request #10044 from dimagi/table-conf-generation

Generate ExportDataSchema for new exports

esoergel committed:
Merge pull request #10055 from dimagi/mobile-dashboard

remove dashboard tab from mobile users unless necessary

NoahCarnahan committed:
Adjust test to use constant literal

emord committed:
cases use case_name

Commit:
update staging.yaml

[ci skip]

orangejenny committed:
update submodule

dannyroberts committed:
run npm update on fab deploy

dannyroberts committed:
update staging.yaml

[ci skip]

emord committed:
Update dockerhq.sh

@snopoke

dannyroberts committed:
update staging.yaml

[ci skip]

emord committed:
Update README.md

emord committed:
Update bootstrap.sh

snopoke committed:
Merge pull request #10065 from dimagi/emord-patch-1

Update dockerhq.sh

orangejenny committed:
get rid of extra line when hovering over menu icon preview

dannyroberts committed:
switch order and remove npm prune for now

dannyroberts committed:
have fab copy node_modules from current release

dannyroberts committed:
change xpath to js-xpath for npm 1.x compat

Commit:
get_all_app_ids

Commit:
Remove abstract method

Commit:
do not use app_id for case schema generation

dannyroberts committed:
have _npm_install do prune, install, AND update

update alone doesn't seem to install js-xapth on npm 1.x

Commit:
Use case_property_mapping

dannyroberts committed:
add server-side xpath hashtag validation

and test

Commit:
Merge pull request #10063 from dimagi/domain_type

Remove make new commtrack domain feature

esoergel committed:
Add README to couchapps

[ci skip]

dannyroberts committed:
add hashtag validation to client side

dannyroberts committed:
update staging.yaml

[ci skip]

Commit:
Adds a tag to ExportItem and adds case props to history

NoahCarnahan committed:
Merge pull request #10064 from dimagi/fab-npm

run npm update on fab deploy

dannyroberts committed:
update staging.yaml

[ci skip]

dannyroberts committed:
update staging.yaml

[ci skip]

dannyroberts committed:
update staging.yaml

[ci skip]

dannyroberts committed:
pin bower js-xpath to old version

emord committed:
Merge pull request #10067 from dimagi/fix-hashtag-validation

Fix hashtag validation

dannyroberts committed:
make form filter allow_case_hashtags server-side

emord committed:
Merge pull request #10068 from dimagi/fix-hashtag-validation

make form filter allow_case_hashtags server-side

NoahCarnahan committed:
Update translations.

[ci skip]

Commit:
update staging.yaml

[ci skip]

Commit:
Refactor backend map to use class-based view, boostrap3, and postgres models

Commit:
Remove unneeded signal code

Commit:
Remove more unused templates

esoergel committed:
Doctype migrations README updates

esoergel committed:
Register new db and migrator

esoergel committed:
Register appropriate couchapps views to new db

Notably, copy `cloudcare_apps` into couchapps, which seems like a better
home for it.  My plan is to only register this to the new db, and then
switch the calling code to this view in the flip PR.

Note that this change shouldn't result in any new views being indexed on
the main db (as I understand it, anyways).

Commit:
Remove setting Domain default_sms_backend_id directly

Commit:
Remove reference to signals code

Commit:
Add warning message to global backend delete

Commit:
Lint

esoergel committed:
Make example a little more elaborate

Not sure how crazy I am about this change - it might be too in-depth and
detract from the main point of the documentation.  Thoughts?
[ci skip]

millerdev committed:
Merge pull request #10062 from dimagi/name-to-case_name

cases use case_name

Commit:
Fix up tests

proteusvacuum committed:
Merge pull request #10026 from dimagi/move-sms-backend-code

Make Gateways use Postgres Models

proteusvacuum committed:
Merge pull request #9514 from dimagi/add-tableau-dashboard

Add ICDS tableau dashboard wrapper

czue committed:
don't run every minute

proteusvacuum committed:
Merge pull request #10071 from dimagi/pillow-seq-task

don't run every minute

snopoke committed:
Merge pull request #10058 from dimagi/sql-cases-to-elastic

sync sql case updates to kafka

czue committed:
Merge pull request #10036 from dimagi/mark-password

Mark password

snopoke committed:
remove vendor volume from docker web container

czue committed:
Merge pull request #10070 from dimagi/couchapps-docs

Add README to couchapps

czue committed:
remove demo/test pillows

czue committed:
don't fail hard if no login info

TylerSheffels committed:
Add analytics tracking language to product agreement

czue committed:
[ci skip] use correct superclass

proteusvacuum committed:
Merge pull request #10076 from dimagi/wv-fix

[ci skip] use correct superclass

proteusvacuum committed:
Merge pull request #10074 from dimagi/empty-login

don't fail hard if no login info

kaapstorm committed:
Merge pull request #10075 from dimagi/update-product-agreement

Add analytics tracking language to product agreement

snopoke committed:
print progress

snopoke committed:
use index name

czue committed:
fix bad variable reference

kaapstorm committed:
Offer app-aware sync from HQ Admin phone restore

Commit:
Merge master

czue committed:
failing test for edge cases of the IN operator

czue committed:
be more lenient with bad args

czue committed:
trap all errors on a per-document basis when rebuilding a table

czue committed:
[ci skip] wrap in parens

proteusvacuum committed:
Merge pull request #10073 from dimagi/kill-demo-pillows

remove demo/test pillows

proteusvacuum committed:
Merge pull request #10082 from dimagi/tidy-up

[ci skip] wrap in parens

snopoke committed:
reuse common vars

snopoke committed:
Merge pull request #10079 from dimagi/in-operator

make IN operator better at dealing with errors

snopoke committed:
Merge pull request #10078 from dimagi/bad-ref

fix bad variable reference

czue committed:
Merge pull request #10080 from dimagi/hqadmin_app-aware_restore

Offer app-aware sync from HQ Admin phone restore

czue committed:
Merge pull request #10077 from dimagi/preindex

tweaks

czue committed:
Merge pull request #10059 from lwyszomi/211968-case

Fixed Care Pathways custom reports

czue committed:
Merge pull request #10072 from dimagi/volume

remove vendor volume from docker web container

czue committed:
Merge pull request #10061 from dimagi/switch-test

Test constant literals in switch expressions

snopoke committed:
improvements to reduce size of image

Commit:
Fix use of shared backends

snopoke committed:
create base image for publishing

snopoke committed:
remove duplicate command in bootstrap.sh

Commit:
Update language used

NoahCarnahan committed:
use constant literals

NoahCarnahan committed:
update rwanda target

Commit:
Fix tests

Commit:
Make use of test backend more consistent across tests

Commit:
add doc strings

Commit:
DocumentSchema -> Document

dannyroberts committed:
fix some error messages

Commit:
repeat_path -> path for consistency

Commit:
add app label

czue committed:
don't fail hard accessing lastest change id from kafka if nothing there

czue committed:
even more logging

czue committed:
don't rebuild tables more than once

Commit:
mobile UCR username filter should use the raw username

millerdev committed:
Move DEFAULT_BUCKET TO corehq.blobs module

millerdev committed:
Add a few missing tests for s3db

millerdev committed:
Implement blobdb for migrating from FS to S3

Commit:
Merge pull request #10084 from dimagi/abt-reports

Update abt report target number

Commit:
create_default_from_item

Commit:
appVersion -> app_version

emord committed:
Merge pull request #10085 from dimagi/respond-to-10067-feedback

fix some error messages

Commit:
vellum:efec854211f4f8657167dee3533d960c6131429c

dannyroberts committed:
update staging.yaml

[ci skip]

Commit:
tests for ExportInstanceGeneration

Commit:
default show and selected to False

Commit:
repeat_path -> path

Commit:
fix tests

snopoke committed:
Merge pull request #10066 from dimagi/case-schema-generation

Case schema generation

dannyroberts committed:
update staging.yaml

[ci skip]

snopoke committed:
update rebuild to build base image first

snopoke committed:
Merge branch 'master' into vars

Commit:
apply paragraph formatting

Commit:
pass credits context to email

Commit:
remove awkward space in syntax

Commit:
add text to email template

Commit:
vellum:5f832fe

snopoke committed:
don't crash in __repr__ for DefaultPreindexPlugin

snopoke committed:
only run preindex plugins for 'default' database

snopoke committed:
sharding tests run on docker

esoergel committed:
Get rid of cloudcare ddoc shenanigans

esoergel committed:
Add new db to test

esoergel committed:
Elaborate on set-up phase (this is getting long)

esoergel committed:
Sync cloudcare and app_manager views to new db

dannyroberts committed:
Merge pull request #10092 from dimagi/sync-default

only run preindex plugins for 'default' database

millerdev committed:
Make generic blob db adaptor for migrations

millerdev committed:
Fix test failing when run against Riak CS

millerdev committed:
PEP8

dimagimon commented:
There are 24 errors:

* corehq/apps/app_manager/detail_screen.py, line 282 - E501 line too long (123 > 115 characters)
* corehq/apps/app_manager/tests/test_case_detail_distance.py, line 31 - E501 line too long (151 > 115 characters)
* corehq/apps/app_manager/tests/test_case_detail_distance.py, line 70 - E501 line too long (151 > 115 characters)
* corehq/apps/app_manager/xpath_validator/tests.py, line 15 - E251 unexpected spaces around keyword / parameter equals
* corehq/apps/app_manager/xpath_validator/tests.py, line 24 - E501 line too long (188 > 115 characters)
* corehq/apps/app_manager/xpath_validator/tests.py, line 30 - E501 line too long (191 > 115 characters)
* corehq/apps/app_manager/xpath_validator/tests.py, line 31 - E251 unexpected spaces around keyword / parameter equals
* corehq/apps/commtrack/tests/test_locations.py, line 118 - F821 undefined name 'make_supply_point'
* corehq/apps/export/models/__init__.py, line 1 - F401 'FormQuestionSchema' imported but unused
* corehq/apps/export/models/__init__.py, line 1 - F401 'QuestionMeta' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'CaseExportDataSchema' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ExportColumn' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ExportDataSchema' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ExportGroupSchema' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ExportInstance' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ExportItem' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ExportRow' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'FormExportDataSchema' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'MultipleChoiceItem' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'Option' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ScalarItem' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'TableConfiguration' imported but unused
* corehq/apps/export/models/new.py, line 103 - E501 line too long (126 > 115 characters)
* corehq/apps/hqadmin/reports.py, line 682 - E501 line too long (116 > 115 characters)

dimagimon commented:
There are 24 errors:

* corehq/apps/app_manager/detail_screen.py, line 282 - E501 line too long (123 > 115 characters)
* corehq/apps/app_manager/tests/test_case_detail_distance.py, line 31 - E501 line too long (151 > 115 characters)
* corehq/apps/app_manager/tests/test_case_detail_distance.py, line 70 - E501 line too long (151 > 115 characters)
* corehq/apps/app_manager/xpath_validator/tests.py, line 15 - E251 unexpected spaces around keyword / parameter equals
* corehq/apps/app_manager/xpath_validator/tests.py, line 24 - E501 line too long (188 > 115 characters)
* corehq/apps/app_manager/xpath_validator/tests.py, line 30 - E501 line too long (191 > 115 characters)
* corehq/apps/app_manager/xpath_validator/tests.py, line 31 - E251 unexpected spaces around keyword / parameter equals
* corehq/apps/commtrack/tests/test_locations.py, line 118 - F821 undefined name 'make_supply_point'
* corehq/apps/export/models/__init__.py, line 1 - F401 'FormQuestionSchema' imported but unused
* corehq/apps/export/models/__init__.py, line 1 - F401 'QuestionMeta' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'CaseExportDataSchema' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ExportColumn' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ExportDataSchema' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ExportGroupSchema' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ExportInstance' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ExportItem' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ExportRow' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'FormExportDataSchema' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'MultipleChoiceItem' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'Option' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ScalarItem' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'TableConfiguration' imported but unused
* corehq/apps/export/models/new.py, line 103 - E501 line too long (126 > 115 characters)
* corehq/apps/hqadmin/reports.py, line 682 - E501 line too long (116 > 115 characters)

esoergel committed:
update staging.yaml

[ci skip]

dannyroberts committed:
Merge pull request #10069 from dimagi/migrate-apps

Register apps database migration

dannyroberts committed:
Merge pull request #10091 from dimagi/credit-line-email

Add remaining credits to invoice email

esoergel committed:
Redirect on successful form submission

millerdev committed:
Merge pull request #10093 from dimagi/docker-sharded-tests

sharding tests run on docker

benrudolph committed:
update staging.yaml

[ci skip]

proteusvacuum committed:
Merge pull request #10086 from dimagi/pillow-api

don't fail hard accessing lastest change id from kafka if nothing there

czue committed:
Merge pull request #10088 from dimagi/filter-by-raw_username

mobile UCR username filter should use the raw username

czue committed:
[ci skip] skip more pact tests

proteusvacuum committed:
Merge pull request #10096 from dimagi/pact-tests

[ci skip] skip more pact tests

proteusvacuum committed:
Merge pull request #10081 from dimagi/more-general-catch

trap all errors on a per-document basis when rebuilding a table

czue committed:
make get_tables_to_rebuild return a set

czue committed:
Merge pull request #10095 from dimagi/redirect-on-success

Redirect on successful form submission

snopoke committed:
Merge pull request #10090 from dimagi/export-instance-generation

Export instance generation

snopoke committed:
Merge pull request #10083 from dimagi/vars

docker improvements

proteusvacuum committed:
Merge pull request #10087 from dimagi/more-logging

don't rebuild tables more than once + logging

czue committed:
Merge pull request #9558 from dimagi/use_e

Use lxml's E instead of xml's Element

NoahCarnahan committed:
Convert remaining constants to literals

dannyroberts committed:
Merge pull request #10094 from dimagi/fsdb2s3db

C3PO and R2D2 went for a walk

NoahCarnahan committed:
fix typos

NoahCarnahan committed:
condense property paths

NoahCarnahan committed:
Remove unnecessary datatypes

NoahCarnahan committed:
Add tanzania targets

Commit:
Merge pull request #10101 from dimagi/abt-reports

Add Tanzania target numbers to Abt reports

NoahCarnahan committed:
Add missing tour analytics event

Commit:
Merge pull request #10105 from dimagi/tour-analytics

Add missing tour analytics event

Commit:
Clean up logic

snopoke committed:
link Dockerfile

snopoke committed:
Merge pull request #10106 from dimagi/docker-link

link Dockerfile

esoergel committed:
Merge pull request #10057 from dimagi/update-view-backend-refs

Make backend views use postgres models

esoergel committed:
Fix bad merge

dimagimon commented:
There are 23 errors:

* corehq/apps/app_manager/detail_screen.py, line 282 - E501 line too long (123 > 115 characters)
* corehq/apps/app_manager/tests/test_case_detail_distance.py, line 31 - E501 line too long (151 > 115 characters)
* corehq/apps/app_manager/tests/test_case_detail_distance.py, line 70 - E501 line too long (151 > 115 characters)
* corehq/apps/app_manager/xpath_validator/tests.py, line 15 - E251 unexpected spaces around keyword / parameter equals
* corehq/apps/app_manager/xpath_validator/tests.py, line 24 - E501 line too long (188 > 115 characters)
* corehq/apps/app_manager/xpath_validator/tests.py, line 30 - E501 line too long (191 > 115 characters)
* corehq/apps/app_manager/xpath_validator/tests.py, line 31 - E251 unexpected spaces around keyword / parameter equals
* corehq/apps/export/models/__init__.py, line 1 - F401 'FormQuestionSchema' imported but unused
* corehq/apps/export/models/__init__.py, line 1 - F401 'QuestionMeta' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'CaseExportDataSchema' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ExportColumn' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ExportDataSchema' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ExportGroupSchema' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ExportInstance' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ExportItem' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ExportRow' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'FormExportDataSchema' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'MultipleChoiceItem' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'Option' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ScalarItem' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'TableConfiguration' imported but unused
* corehq/apps/export/models/new.py, line 103 - E501 line too long (126 > 115 characters)
* corehq/apps/hqadmin/reports.py, line 682 - E501 line too long (116 > 115 characters)

dimagimon commented:
There are 23 errors:

* corehq/apps/app_manager/detail_screen.py, line 282 - E501 line too long (123 > 115 characters)
* corehq/apps/app_manager/tests/test_case_detail_distance.py, line 31 - E501 line too long (151 > 115 characters)
* corehq/apps/app_manager/tests/test_case_detail_distance.py, line 70 - E501 line too long (151 > 115 characters)
* corehq/apps/app_manager/xpath_validator/tests.py, line 15 - E251 unexpected spaces around keyword / parameter equals
* corehq/apps/app_manager/xpath_validator/tests.py, line 24 - E501 line too long (188 > 115 characters)
* corehq/apps/app_manager/xpath_validator/tests.py, line 30 - E501 line too long (191 > 115 characters)
* corehq/apps/app_manager/xpath_validator/tests.py, line 31 - E251 unexpected spaces around keyword / parameter equals
* corehq/apps/export/models/__init__.py, line 1 - F401 'FormQuestionSchema' imported but unused
* corehq/apps/export/models/__init__.py, line 1 - F401 'QuestionMeta' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'CaseExportDataSchema' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ExportColumn' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ExportDataSchema' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ExportGroupSchema' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ExportInstance' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ExportItem' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ExportRow' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'FormExportDataSchema' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'MultipleChoiceItem' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'Option' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ScalarItem' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'TableConfiguration' imported but unused
* corehq/apps/export/models/new.py, line 126 - E501 line too long (126 > 115 characters)
* corehq/apps/hqadmin/reports.py, line 682 - E501 line too long (116 > 115 characters)

dimagimon commented:
There are 23 errors:

* corehq/apps/app_manager/detail_screen.py, line 282 - E501 line too long (123 > 115 characters)
* corehq/apps/app_manager/tests/test_case_detail_distance.py, line 31 - E501 line too long (151 > 115 characters)
* corehq/apps/app_manager/tests/test_case_detail_distance.py, line 70 - E501 line too long (151 > 115 characters)
* corehq/apps/app_manager/xpath_validator/tests.py, line 15 - E251 unexpected spaces around keyword / parameter equals
* corehq/apps/app_manager/xpath_validator/tests.py, line 24 - E501 line too long (188 > 115 characters)
* corehq/apps/app_manager/xpath_validator/tests.py, line 30 - E501 line too long (191 > 115 characters)
* corehq/apps/app_manager/xpath_validator/tests.py, line 31 - E251 unexpected spaces around keyword / parameter equals
* corehq/apps/export/models/__init__.py, line 1 - F401 'FormQuestionSchema' imported but unused
* corehq/apps/export/models/__init__.py, line 1 - F401 'QuestionMeta' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'CaseExportDataSchema' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ExportColumn' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ExportDataSchema' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ExportGroupSchema' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ExportInstance' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ExportItem' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ExportRow' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'FormExportDataSchema' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'MultipleChoiceItem' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'Option' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ScalarItem' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'TableConfiguration' imported but unused
* corehq/apps/export/models/new.py, line 126 - E501 line too long (126 > 115 characters)
* corehq/apps/hqadmin/reports.py, line 682 - E501 line too long (116 > 115 characters)

esoergel commented:
Closing to merge https://github.com/dimagi/commcare-hq/pull/9946 directly instead
